azurecosmos/src/main/java/com/yahoo/ycsb/db/AzureCosmosClient.java

```diff
@@ -0,0 +1,454 @@
+                      + " useSinglePartitionCollection={}, useUpsert={}, useHashQueryForScan={}, "
+                      + "queryText={}",
+              uri,
+              connectionPolicy.getConnectionMode(),
+              consistencyLevel.toString(),
+              connectionPolicy.getRetryOptions().getMaxRetryAttemptsOnThrottledRequests(),
+              connectionPolicy.getRetryOptions().getMaxRetryWaitTimeInSeconds(),
+              this.useSinglePartitionCollection,
+              this.useUpsert,
+              this.useHashQueryForScan,
+              this.queryText);
+      AzureCosmosClient.client = new DocumentClient(uri, primaryKey, connectionPolicy, consistencyLevel);
+      LOGGER.info("Azure Cosmos connection created: {}", uri);
+    } catch (Exception e) {
+      throw new DBException("Could not initialize Azure Cosmos client", e);
+    }
+
+    // Verify the database exists
+    try {
+      AzureCosmosClient.client.readDatabase(getDatabaseLink(this.databaseName), new RequestOptions());
+    } catch (Exception e) {
+      throw new DBException("Could not read the database: " + this.databaseName, e);
+    }
+  }
+
+  private String getStringProperty(String propertyName, String defaultValue) {
+    return getProperties().getProperty(propertyName, defaultValue);
+  }
+
+  private boolean getBooleanProperty(String propertyName, boolean defaultValue) {
+    String stringVal = getProperties().getProperty(propertyName, null);
+    if (stringVal == null) {
+      return defaultValue;
+    }
+    return Boolean.parseBoolean(stringVal);
+  }
+
+  private int getIntProperty(String propertyName, int defaultValue) {
+    String stringVal = getProperties().getProperty(propertyName, null);
+    if (stringVal == null) {
+      return defaultValue;
+    }
+    try {
+      return Integer.parseInt(stringVal);
+    } catch (NumberFormatException e) {
+      return defaultValue;
+    }
+  }
+
+  private String getQueryText() {
+    return this.useHashQueryForScan ?
+             "SELECT * FROM root r WHERE r.id = @startkey" :
+             "SELECT TOP @recordcount * FROM root r WHERE r.myid >= @startkey";
+  }
+
+  /**
+   * Cleanup any state for this DB. Called once per DB instance; there is one DB
+   * instance per client thread.
+   */
+  @Override
+  public void cleanup() throws DBException {
+    if (INIT_COUNT.decrementAndGet() == 0) {
+      try {
+        AzureCosmosClient.client.close();
+      } catch (Exception e) {
+        if (!this.includeExceptionStackInLog) {
+          e = null;
+        }
+        LOGGER.error("Could not close DocumentClient", e);
+      } finally {
+        client = null;
+      }
+    }
+  }
+
+  @Override
+  public Status read(String table, String key, Set<String> fields, Map<String, ByteIterator> result) {
+    String documentLink = getDocumentLink(this.databaseName, table, key);
+
+    ResourceResponse<Document> readResource = null;
+    Document document;
+    long startTime = System.nanoTime();
```

Comment:
here and below, are we expecting nano-second level performance for this particular workload? :)) If not, I'd stay with microsecond for usability, but it's up to you.

Reply:
We only work in nano seconds :)